src/dict.c

```diff
@@ -123,133 +130,163 @@ int _dictInit(dict *d, dictType *type)
  * but with the invariant of a USED/BUCKETS ratio near to <= 1 */
-int dictResize(dict *d)
-{
+int dictResize(dict *d) {
     unsigned long minimal;
 
     if (!dict_can_resize || dictIsRehashing(d)) return DICT_ERR;
-    minimal = d->ht_used[0];
+    minimal = d->ht_used;
     if (minimal < DICT_HT_INITIAL_SIZE)
         minimal = DICT_HT_INITIAL_SIZE;
-    return dictExpand(d, minimal);
+    return _dictResize(d, minimal, NULL);
 }
 
-/* Expand or create the hash table,
- * when malloc_failed is non-NULL, it'll avoid panic if malloc fails (in which case it'll be set to 1).
- * Returns DICT_OK if expand was performed, and DICT_ERR if skipped. */
-int _dictExpand(dict *d, unsigned long size, int* malloc_failed)
-{
+/* Resize or create the hash table. When malloc_failed is non-NULL, it'll avoid
+ * panic if malloc fails (in which case it'll be set to 1).
+ * Returns DICT_OK if resize was performed, and DICT_ERR if skipped. */
+int _dictResize(dict *d, unsigned long size, int *malloc_failed) {
     if (malloc_failed) *malloc_failed = 0;
 
     /* the size is invalid if it is smaller than the number of
      * elements already inside the hash table */
-    if (dictIsRehashing(d) || d->ht_used[0] > size)
+    if (dictIsRehashing(d) || d->ht_used > size)
         return DICT_ERR;
 
-    /* the new hash table */
+    /* the new reallocated hash table */
     dictEntry **new_ht_table;
-    unsigned long new_ht_used;
     signed char new_ht_size_exp = _dictNextExp(size);
 
     /* Detect overflows */
-    size_t newsize = 1ul<<new_ht_size_exp;
-    if (newsize < size || newsize * sizeof(dictEntry*) < newsize)
+    size_t newsize = 1ul << new_ht_size_exp;
+    if (newsize < size || newsize * sizeof(dictEntry *) < newsize)
         return DICT_ERR;
 
     /* Rehashing to the same table size is not useful. */
-    if (new_ht_size_exp == d->ht_size_exp[0]) return DICT_ERR;
-
-    /* Allocate the new hash table and initialize all pointers to NULL */
-    if (malloc_failed) {
-        new_ht_table = ztrycalloc(newsize*sizeof(dictEntry*));
-        *malloc_failed = new_ht_table == NULL;
-        if (*malloc_failed)
-            return DICT_ERR;
-    } else
-        new_ht_table = zcalloc(newsize*sizeof(dictEntry*));
-
-    new_ht_used = 0;
+    if (new_ht_size_exp == d->ht_size_exp) return DICT_ERR;
+
+    /* If hash table grows - then realloc() bigger memory. */
+    /* If first init - then this condition still TRUE. realloc() memory */
+    if (d->ht_size_exp < new_ht_size_exp) {
+        /* Realloc the hash table (without initializing new buckets to NULL) */
+        if (malloc_failed) {
+            new_ht_table = ztryrealloc(d->ht_table,
+                                       newsize * sizeof(dictEntry *));
+            *malloc_failed = new_ht_table == NULL;
+            if (*malloc_failed)
+                return DICT_ERR;
+        } else
+            new_ht_table = zrealloc(d->ht_table, newsize * sizeof(dictEntry *));
```

Comment:
Just to cover my understanding. I investigated this awhile ago and found this realloc was slow to complete because jemalloc may do a memcpy of this block of memory which can be slow. (See, https://github.com/redis/redis/issues/9517#issuecomment-923044192) So even though the rest of the operation was similar speed, it seemed like the wrong tradeoff for Redis. Would be good to understand how long this can take, especially since in many workloads a lot of the memory usage in Redis is this hash table.

Reply:
@madolson you are right.
I probably dived into this task too fast and took for granted too many assumptions.  I actually suggested here another optimization of avoiding from memset the memory on allocation (and smaller amount of metadata per instance...), but then again if we use calloc() then we already benefit from kernel optimization of lazy memset ([Why malloc+memset is slower than calloc?](https://stackoverflow.com/questions/2688466/why-mallocmemset-is-slower-than-calloc) ).

Here is quick test on my lazy macos (Optimization -O2)
```
                      zmalloc    zcalloc       mmst zmalc+mmst   zrealloc 
   1048576 bytes:          0          0          0          1          0
   2097152 bytes:          0          0          1          2          2
   4194304 bytes:          1          0          3          2          2
   8388608 bytes:          1          0          4          5          7
  16777216 bytes:          1          1          8         10         11
  33554432 bytes:          2          1         16         20         20
  67108864 bytes:          4          4         33         37         40
 134217728 bytes:          9          8        114         72         79
 268435456 bytes:         16         16        132        140        158
 536870912 bytes:         36         34        246        279        386
```
Obviously there is tradeoff between memory optimization and speed. Either pay the price of having two hash tables or suffer from performance penalty of realloc(). I searched for a quick win, but maybe we should  go back to the drawing board and think more seriously about major change to hashtable - or just say that we are fine :)

Reply:
@moticless please avoid doing any benchmarks (and drawing conclusions) from MacOS runs, this is NOT our target platform (not the same kernel, and not the same allocator).

Reply:
@oranagra, sorry for my bad habbit.

Here numbers for ubuntu container, hosted by macos, Optimization: jemalloc -O3, main -O0

```
                      zmalloc    zcalloc       mmst zmalc+mmst   zrealloc
   1048576 bytes:          0          0          0          0          0
   2097152 bytes:          0          0          0          0          0
   4194304 bytes:          0          0          0          1          0
   8388608 bytes:          0          0          4          3          3
  16777216 bytes:          1          0          5          6          3
  33554432 bytes:          0          0         12         12          6
  67108864 bytes:          0          0         24         23         14
 134217728 bytes:          0          0         46         46         27
 268435456 bytes:          0          0         85         87         52
 536870912 bytes:          0          0        159        178         97
```

Reply:
let's try to think of an extreme case of 1m elements hash table, reallocating 800mb will induce a latency spike of 100ms?
for an extreme case, maybe that's acceptable?
i remind us that before redis 6.2, that would have also started a blocking eviction process, that would have cause a latency spike too.

Reply:
@oranagra Also the reverse is true, now that we're no longer impacted by the blocking eviction spike, we may not want to re-introduce a new delay, especially if we have plenty of memory to spare.